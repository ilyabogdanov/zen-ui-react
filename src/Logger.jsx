
const LEVEL = {
    OFF     : 0,
    WARN    : 1,
    INFO    : 2,
    DEBUG   : 3,
    TRACE   : 4
};
const Logger = {
    level: LEVEL.TRACE,
    error (e) {
        throw new Error(e);
    },
    warn () {
        if (this.level >= LEVEL.WARN) {
            try {
                console.warn.apply(this, ...arguments)
            } catch (e) {
                console.warn.apply(this, arguments)
            }
        }
    },
    info () {
        if (this.level >= LEVEL.INFO) {
            try {
                console.warn.apply(this, ...arguments)
            } catch (e) {
                console.warn.apply(this, arguments)
            }
        }
    },
    debug () {
        if (this.level >= LEVEL.DEBUG) {
            try {
                console.log.apply(this, ...arguments)
            } catch (e) {
                console.log.apply(this, arguments)
            }
        }
    },
    trace () {
        if (this.level >= LEVEL.TRACE) {
            try {
                console.log.apply(this, ...arguments);
            } catch (e) {
                console.log.apply(this, arguments)
            }
        }
    }
};
export default Logger;
export { LEVEL };
