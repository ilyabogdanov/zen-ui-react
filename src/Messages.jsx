
const _VALID_BOOLEAN_VALUE_ = "either true, false, or empty attribute which equals true";
const _VALID_PIXEL_VALUE_ = "either (a) Natural number, starting with anything but zero, ending with \"px\", or (b) Zero, without \"px\"";
const _BAD_HEIGHT_ = "\"height\" must be either (1) undefined, (2) 100%, or (3) valid pixel value, which means "+_VALID_PIXEL_VALUE_+".";
const _READONLY_IS_DISABLED_ = "will be forced to disabled because it is readonly.";

const COMMON_MESSAGES = {
    BAD_PADDING: "Property \"padding\" must be either 1,2 or 4 values, each of which must be valid pixel value, which means "+_VALID_PIXEL_VALUE_+"."
};
const TEXT_MESSAGES = {
    BAD_CLIPPED: "Property \"clipped\" must be but is not valid boolean value, which means "+_VALID_BOOLEAN_VALUE_+"."
};
const CONJOINED_BUTTONS_MESSAGES = {
    INVALID_CHILD: "Conjoined buttons container may contain only these components: push button, link button, toggle button, toggle radio button.",
    NO_CHILDREN: "Conjoined buttons container must have children"
};
const CHECKBOX_MESSAGES = {
    READONLY_IS_DISABLED: "Checkbox "+_READONLY_IS_DISABLED_
};
const RADIO_MESSAGES = {
    READONLY_IS_DISABLED: "Radio "+_READONLY_IS_DISABLED_,
    INVALID_CHILD: "Radio group may contain only Radio component or arrays of them"
};
const HORIZONTAL_LAYOUT_MESSAGES = {
    BAD_HEIGHT: "Horizontal layout's "+_BAD_HEIGHT_,
    ILLEGAL_CHILD: "Horizontal layout may have only columns or arrays of them.",
    NO_CHILDREN: "Horizontal layout must have at least one column.",
    STRETCH_WITH_HEIGHT: "Horizontal layout must not have both \"stretch\" and \"height\" properties.",
    EXCESSIVE_WIDTH: "Total width of all columns must not be more than 100%;",
    INSUFFICIENT_WIDTH: "Either (a) total width of all columns must be 100%, or (b) if less, then ratio may be used to divide remaining space.",
    NO_SPACE_FOR_RATIO: "No free space to accommodate columns with ratio width.",
    INVALID_WIDTH: "Horizontal layout column \"width\" must be either (a) pixel value, (b) percent value, or (c) ratio, which means Natural number.",
    NO_WIDTH: "Horizontal layout column must have \"width\".",
    COLUMN_WIDTH_REQUIRED: "Column with must be set.",
    TOTAL_WIDTH_LIMIT_EXCEEDED: "Total width of all columns, percent and ration, does not equal precisely 100%. It should not cause trouble but be careful anyway.",
    NO_SPACE_FOR_NO_WIDTH: "No free space to accommodate columns with undefined width."
};
const VERTICAL_LAYOUT_MESSAGES = {
    BAD_HEIGHT: "Vertical layout's "+_BAD_HEIGHT_,
    ILLEGAL_CHILD: "Vertical layout may have only rows or arrays of them.",
    NO_CHILDREN: "Vertical layout must have at least one row.",
    MULTIPLE_STRETCH_ROWS: "Vertical layout may have only one stretch row."
};
const WINDOW_MESSAGES = {
    ILLEGAL_CHILD: "Window can't have illegal children",
    NO_CHILDREN: "Window must have head and body"
};
const WINDOW_HEAD_MESSAGES = {
    ILLEGAL_CHILD: "Window head can have only close button"
};
const UTILS_MESSAGES = {
    NO_STYLE_DEST: "style destination must be provided"
};
const TREE_MESSAGES = {
    ILLEGAL_CHILD: "Illegal child, see docs for more info."
};
export {
    COMMON_MESSAGES,
    CONJOINED_BUTTONS_MESSAGES,
    CHECKBOX_MESSAGES,
    RADIO_MESSAGES,
    TEXT_MESSAGES,
    HORIZONTAL_LAYOUT_MESSAGES,
    VERTICAL_LAYOUT_MESSAGES,
    WINDOW_MESSAGES,
    WINDOW_HEAD_MESSAGES,
    TREE_MESSAGES,
    UTILS_MESSAGES
};
