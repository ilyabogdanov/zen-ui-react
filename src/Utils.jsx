import Logger from "./Logger";
import { UTILS_MESSAGES } from "./Messages";

const REGEXP = {
    padding: /^(0|[1-9]\d*px)$|^((0|[1-9]\d*px)\s(0|[1-9]\d*px))$|^(((0|[1-9]\d*px)\s){3}(0|[1-9]\d*px))$/i,
    pixelValue: /^([1-9](\d*))(px)$/i,
    remValue: /^((0|[1-9](\d*)))(\.[0-9]+)?(rem)$/i,
    percentValue: /^((0|[1-9](\d*)))(\.[0-9]+)?(%)$/i,
    ratioValue: /^([1-9](\d*))$/i,
}
const SAFE_STYLE = [
    "opacity",
    "background",
    "whiteSpace"
];
const Utils = {
    id: 0,
    idPrefix: "zen-ui-id-",
    generateId: function (e) {
        return this.idPrefix+(++this.id);
    },
    isArray: function (e) {
        return Object.prototype.toString.call(e) === "[object Array]";
    },
    isDefined: function (e) {
        return typeof e !== "undefined";
    },
    isUndefined: function (e) {
        return typeof e === "undefined";
    },
    isValidPadding: function (e) {
        return REGEXP.padding.test(e);
    },
    isPercentValue: function (e) {
        return REGEXP.percentValue.test(e);
    },
    isRemValue: function (e) {
        return REGEXP.remValue.test(e);
    },
    isPixelValue: function (e) {
        return REGEXP.pixelValue.test(e);
    },
    isRatioValue: function (e) {
        return REGEXP.ratioValue.test(e);
    },
    isZero: function (e) {
        return e===0||e==="0";
    },
    getPercentValue: function (e) {
        let result = REGEXP.percentValue.exec(e);
        if (!result) {
            return null;
        }
        if (typeof result[4]==="undefined") {
            return parseInt(result[1]+result[5]);
        } else {
            return parseFloat(result[1]+result[4]+result[5]);
        }
    },
    isSafeStyle: function (e) {
        return typeof e === "string" && SAFE_STYLE.indexOf(e) !== -1;
    },
    setSafeStyle: function (src, dest) {
        if (!src || src === {} || typeof src !== "object") {
            // silence
        } else if (!dest || typeof dest !== "object") {
            Logger.error(UTILS_MESSAGES.NO_STYLE_DEST);
        } else {
            for (var e in src) {
                if (this.isSafeStyle(e)) {
                    dest[e] = src[e];
                }
            }
        }
    }
};
export default Utils;
