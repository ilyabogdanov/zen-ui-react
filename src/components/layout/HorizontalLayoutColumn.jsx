import React from "react";
import Utils from "../../Utils";

/**
 * @class
 * @classdesc Horizontal layout column
 * @author Ilya Bogdanov <public@ilyabogdanov.ru>
 * @license Apache-2.0
 *
 * @prop {string} props.padding
 * @prop {string} props.width
 * @prop {object} props.style
 *
 * @prop {*} props.children
 */
export default class HorizontalLayoutColumn extends React.Component {
    render() {
        const { width, padding, children, style } = this.props;

        const percentWidth = Utils.isDefined(width) && Utils.isPercentValue(width);
        const fixedWidth = Utils.isDefined(width) && (Utils.isPixelValue(width) || Utils.isRemValue(width));

        let outerStyle = percentWidth ? { width: width } : {};
        Utils.setSafeStyle(style, outerStyle);

        return (
            <div className="zen_ui__horizontal_layout_column" style={outerStyle}>
                <div style={{ padding: padding ? padding : null, width: fixedWidth ? width : null }}>
                    <div>
                        {children}
                    </div>
                </div>
            </div>
        );
    }
};
