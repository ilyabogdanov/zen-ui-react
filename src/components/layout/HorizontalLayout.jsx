import React from "react";
import Logger from "../../Logger";
import Utils from "../../Utils";
import { COMMON_MESSAGES, HORIZONTAL_LAYOUT_MESSAGES } from "../../Messages";
import HorizontalLayoutColumn from "./HorizontalLayoutColumn";

/**
 * @class
 * @classdesc Horizontal layout
 * @author Ilya Bogdanov <public@ilyabogdanov.ru>
 * @license Apache-2.0
 *
 * @prop {string} props.height
 * @prop {boolean} props.divider
 * @prop {object} props.style
 * @prop {*} props.children
 */
export default class HorizontalLayout extends React.Component {
    render() {
        const { children, height, divider, style } = this.props;
        const { name } = this.constructor;

        let columnWidth = {
            percent: 0,
            ratio: 0,
            none: 0
        };
        let percentage = {
            initial: 0,
            used: 0,
            remaining: 100
        };
        let referenceUnit = null;
        
        if (!children) {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.NO_CHILDREN);
        }
        if (Utils.isDefined(height) && !Utils.isZero(height) && !Utils.isPixelValue(height) && !Utils.isRemValue(height) && height!=="100%") {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.BAD_HEIGHT);
        }

        checkChildren(children);

        function checkChildren (e) {

            if (Utils.isArray(e)) {
                e.forEach(e => checkChildren(e));

            } else if (!e || !e.type || e.type !== HorizontalLayoutColumn) {
                Logger.error(HORIZONTAL_LAYOUT_MESSAGES.ILLEGAL_CHILD);

            } else {

                if (Utils.isDefined(e.props.padding) && !Utils.isValidPadding(e.props.padding)) {
                    Logger.error(COMMON_MESSAGES.BAD_PADDING);
                }
                if (Utils.isDefined(e.props.width)) {

                    if (typeof e.props.width === "boolean") {
                        Logger.error(HORIZONTAL_LAYOUT_MESSAGES.INVALID_WIDTH);

                    } else if (Utils.isPercentValue(e.props.width)) {
                        columnWidth.percent += Utils.getPercentValue(e.props.width);

                    } else if (Utils.isRatioValue(e.props.width)) {
                        columnWidth.ratio += parseInt(e.props.width);

                    } else if (Utils.isPixelValue(e.props.width) || Utils.isRemValue(e.props.width)) {
                        // NOP

                    } else {
                        Logger.error(HORIZONTAL_LAYOUT_MESSAGES.INVALID_WIDTH);
                    }
                } else {
                    columnWidth.none++;
                }
            }
        }

        percentage.initial = columnWidth.percent;
        percentage.remaining -= percentage.initial;
        Logger.trace(name, "percentage:", percentage);

        if (percentage.initial > 100) {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.EXCESSIVE_WIDTH);
            
        } else if (percentage.initial >= 0 && percentage.initial < 100 && columnWidth.ratio === 0/* && columnWidth.none===0*/) {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.INSUFFICIENT_WIDTH);
            
        } else if (percentage.initial === 100 && columnWidth.ratio > 0) {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.NO_SPACE_FOR_RATIO);
            
        } /* else if ((percentage.initial === 100 && columnWidth.none > 0) || (columnWidth.ratio > 0 && columnWidth.none > 0)) {
            Logger.error(HORIZONTAL_LAYOUT_MESSAGES.NO_SPACE_FOR_NO_WIDTH);
            
        }*/ else if (percentage.initial < 100 && columnWidth.ratio > 0) {
            referenceUnit = Math.round((percentage.remaining / columnWidth.ratio) * 100) / 100;

        }/* else if (percentage.initial < 100 && columnWidth.none > 0) {
            referenceUnit = Math.round((percentage.remaining / columnWidth.none) * 100) / 100;
        }*/
        referenceUnit && Logger.trace(name, "reference unit:", referenceUnit+"%");

        const updatedChildren = !referenceUnit ? null : React.Children.map(children, (e) => {
            if (Utils.isRatioValue(e.props.width)) {
                let next;
                next = Math.round(referenceUnit * e.props.width * 100) / 100;
                Logger.trace(name, "used", percentage.used, "+", "next", next, "=", (Math.round((percentage.used + next) * 100) / 100));
                percentage.used = Math.round((percentage.used + next) * 100) / 100;
                return React.cloneElement(e, {width: next+"%"});
            } else {
                return React.cloneElement(e, {});
            }
        });

        if (percentage.initial !== 100) {
            if (percentage.used !== percentage.remaining) {
                Logger.warn(HORIZONTAL_LAYOUT_MESSAGES.TOTAL_WIDTH_LIMIT_EXCEEDED);
            } else {
                Logger.trace(name, "accurate result");
            }
        }

        let outerStyle = height ? { height: height } : {};
        Utils.setSafeStyle(style, outerStyle);

        return (
            <div className="zen_ui__horizontal_layout" style={outerStyle} data-divider={divider}>
                {referenceUnit ? updatedChildren : children}
            </div>
        );
    }
};
