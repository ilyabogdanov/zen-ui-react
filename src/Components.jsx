import React from "react";

/**
 * button
 */
import Button_Props           from "./components/form/Button_Props";
import Combobox               from "./components/form/Combobox";
import ConjoinedButtons       from "./components/form/ConjoinedButtons";
import LinkButton             from "./components/form/LinkButton";
import PushButton             from "./components/form/PushButton";
import ToggleButton           from "./components/form/ToggleButton";
import ToggleRadioButton            from "./components/form/ToggleRadioButton";

/**
 * content
 */
import Text                   from "./components/content/Text";
import Text_Props             from "./components/content/Text_Props";

/**
 * input
 */
import Checkbox               from "./components/form/Checkbox";
import Checkbox_Props         from "./components/form/Checkbox_Props";
import NumberField            from "./components/form/NumberField";
import PasswordField          from "./components/form/PasswordField";
import Radio                  from "./components/form/Radio";
import TextField              from "./components/form/TextField";
import TimeField              from "./components/form/TimeField";

/**
 * layout
 */
import HorizontalLayout       from "./components/layout/HorizontalLayout";
import HorizontalLayoutColumn from "./components/layout/HorizontalLayoutColumn";
import VerticalLayout         from "./components/layout/VerticalLayout";
import VerticalLayoutRow      from "./components/layout/VerticalLayoutRow";

/**
 * navigation
 */
import HorizontalMenu         from "./components/navigation/HorizontalMenu";
import HorizontalMenuImage    from "./components/navigation/HorizontalMenuImage";
import HorizontalMenuLink     from "./components/navigation/HorizontalMenuLink";
import HorizontalMenuColumn   from "./components/navigation/HorizontalMenuColumn";
import Path                   from "./components/navigation/Path";
import PathItem               from "./components/navigation/PathItem";

/**
 * panel
 */
import LoadingMask            from "./components/panel/LoadingMask";
import Tree                   from "./components/panel/Tree";
import TreeBranch             from "./components/panel/TreeBranch";
import TreeBranchlet          from "./components/panel/TreeBranchlet";
import TreeLeaf               from "./components/panel/TreeLeaf";
import TreeLeafColumn         from "./components/panel/TreeLeafColumn";
import TreeLinkButton         from "./components/panel/TreeLinkButton";
import TreePushButton         from "./components/panel/TreePushButton";

/**
 * window
 */
import Modal                  from "./components/window/Modal";
import Window                 from "./components/window/Window";
import Window_Props           from "./components/window/Window_Props";
import WindowBody             from "./components/window/WindowBody";
import WindowCloseButton      from "./components/window/WindowCloseButton";
import WindowHead             from "./components/window/WindowHead";

/**
 * Container
 */
import ScrollArea                  from "./components/container/ScrollArea";

const Zen = {
    Button_Props: Button_Props,
    Combobox: Combobox,
    ConjoinedButtons: ConjoinedButtons,
    LinkButton: LinkButton,
    PushButton: PushButton,
    ToggleButton: ToggleButton,
    ToggleRadioButton: ToggleRadioButton,
    Text: Text,
    Text_Props: Text_Props,
    Checkbox: Checkbox,
    Checkbox_Props: Checkbox_Props,
    NumberField: NumberField,
    PasswordField: PasswordField,
    Radio: Radio,
    TextField: TextField,
    TimeField: TimeField,
    HorizontalLayout: HorizontalLayout,
    HorizontalLayoutColumn: HorizontalLayoutColumn,
    VerticalLayout: VerticalLayout,
    VerticalLayoutRow: VerticalLayoutRow,
    HorizontalMenu: HorizontalMenu,
    HorizontalMenuImage: HorizontalMenuImage,
    HorizontalMenuLink: HorizontalMenuLink,
    HorizontalMenuColumn: HorizontalMenuColumn,
    Path: Path,
    PathItem: PathItem,
    LoadingMask: LoadingMask,
    Tree: Tree,
    TreeBranch: TreeBranch,
    TreeBranchlet: TreeBranchlet,
    TreeLeaf: TreeLeaf,
    TreeLeafColumn: TreeLeafColumn,
    TreeLinkButton: TreeLinkButton,
    TreePushButton: TreePushButton,
    Modal: Modal,
    Window: Window,
    Window_Props: Window_Props,
    WindowBody: WindowBody,
    WindowCloseButton: WindowCloseButton,
    WindowHead: WindowHead,
    ScrollArea: ScrollArea
};

export {

    // bundle of all components
    Zen,

    // button
    Button_Props,
    Combobox,
    ConjoinedButtons,
    LinkButton,
    PushButton,
    ToggleButton,
    ToggleRadioButton,
    
    // content
    Text,
    Text_Props,

    // input
    Checkbox,
    Checkbox_Props,
    NumberField,
    PasswordField,
    Radio,
    TextField,
    TimeField,

    // layout
    HorizontalLayout,
    HorizontalLayoutColumn,
    VerticalLayout,
    VerticalLayoutRow,

    // navigation
    HorizontalMenu,
    HorizontalMenuImage,
    HorizontalMenuLink,
    HorizontalMenuColumn,
    Path,
    PathItem,

    // panel
    LoadingMask,
    Tree,
    TreeBranch,
    TreeBranchlet,
    TreeLeaf,
    TreeLeafColumn,
    TreeLinkButton,
    TreePushButton,

    // window
    Modal,
    Window,
    Window_Props,
    WindowBody,
    WindowCloseButton,
    WindowHead,

    // container
    ScrollArea
}
