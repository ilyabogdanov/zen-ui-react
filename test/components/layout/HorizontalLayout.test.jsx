import React from "react";
import renderer from "react-test-renderer";
import HorizontalLayout from "../../../src/components/layout/HorizontalLayout";
import HorizontalLayoutColumn from "../../../src/components/layout/HorizontalLayoutColumn";
import { COMMON_MESSAGES, HORIZONTAL_LAYOUT_MESSAGES } from "../../../src/Messages";
import Logger, { LEVEL } from "../../../src/Logger";

// Horizontal layout has log messages with level = TRACE.
// Probably you don't want to see them during test.
Logger.level = LEVEL.OFF;

describe("Horizontal Layout", () => {
    test("without children should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout/>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.NO_CHILDREN);
    });
    test("with illegal children should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <div/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.ILLEGAL_CHILD);
    });
    test("with empty attribute height should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout height>
                    <HorizontalLayoutColumn/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.BAD_HEIGHT);
    });
    test("with invalid height should throw error", () => {
        ["","1","123%","1.23px"].forEach(e => {
            expect(() => {
                renderer.create(
                    <HorizontalLayout height={e}>
                        <HorizontalLayoutColumn/>
                    </HorizontalLayout>
                );
            }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.BAD_HEIGHT);
        });
    });
    test("should render correctly", () => {
        const e1 = renderer.create(
            <HorizontalLayout height="100%">
                <HorizontalLayoutColumn width="10px"/>
                <HorizontalLayoutColumn width="20px" padding="1px 2px 3px 4px"/>
                <HorizontalLayoutColumn width="10%"/>
                <HorizontalLayoutColumn width="20%"/>
                <HorizontalLayoutColumn width="1"/>
                <HorizontalLayoutColumn width="2"/>
            </HorizontalLayout>
        ).toJSON();
        const e2 = renderer.create(
            <HorizontalLayout height="100%">
                <HorizontalLayoutColumn width={1}/>
                <HorizontalLayoutColumn/>
            </HorizontalLayout>
        ).toJSON();
        const e3 = renderer.create(
            <HorizontalLayout height="100%">
                <HorizontalLayoutColumn width="100%"/>
            </HorizontalLayout>
        ).toJSON();
        const e4 = renderer.create(
            <HorizontalLayout>
                <HorizontalLayoutColumn width={1}/>
            </HorizontalLayout>
        ).toJSON();
        expect(e1).toMatchSnapshot();
        expect(e2).toMatchSnapshot();
        expect(e3).toMatchSnapshot();
        expect(e4).toMatchSnapshot();
    });
    test("with unused space should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <HorizontalLayoutColumn width="50%"/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.INSUFFICIENT_WIDTH);
    });
    test("without free space and with ratio-width columns should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <HorizontalLayoutColumn width="100%"/>
                    <HorizontalLayoutColumn width="1"/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.NO_SPACE_FOR_RATIO);
    });
});
describe("Horizontal Layout column", () => {
    test("with empty width should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <HorizontalLayoutColumn width/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.INVALID_WIDTH);
    });
    test("with invalid width should throw error", () => {
        ["","0"].forEach(e => {
            expect(() => {
                renderer.create(
                    <HorizontalLayout>
                        <HorizontalLayoutColumn width={e}/>
                    </HorizontalLayout>
                );
            }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.INVALID_WIDTH);
        });
    });
    test("with greater than 100% width should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <HorizontalLayoutColumn width="123%"/>
                </HorizontalLayout>
            );
        }).toThrowError(HORIZONTAL_LAYOUT_MESSAGES.EXCESSIVE_WIDTH);
    });
    test("with empty padding should throw error", () => {
        expect(() => {
            renderer.create(
                <HorizontalLayout>
                    <HorizontalLayoutColumn padding/>
                </HorizontalLayout>
            );
        }).toThrowError(COMMON_MESSAGES.BAD_PADDING);
    });
    test("with invalid padding should throw error", () => {
        ["","1","123%","1.23px"].forEach(e => {
            expect(() => {
                renderer.create(
                    <HorizontalLayout>
                        <HorizontalLayoutColumn padding={e}/>
                    </HorizontalLayout>
                );
            }).toThrowError(COMMON_MESSAGES.BAD_PADDING);
        });
    });
});
