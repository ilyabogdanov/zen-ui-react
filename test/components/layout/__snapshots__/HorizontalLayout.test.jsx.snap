// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Horizontal Layout should render correctly 1`] = `
<div
  className="zen_ui__horizontal_layout"
  data-divider={undefined}
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="zen_ui__horizontal_layout_column"
    style={Object {}}
  >
    <div
      style={
        Object {
          "padding": null,
          "width": "10px",
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={Object {}}
  >
    <div
      style={
        Object {
          "padding": "1px 2px 3px 4px",
          "width": "20px",
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "10%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "20%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "23.33%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "46.66%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Horizontal Layout should render correctly 2`] = `
<div
  className="zen_ui__horizontal_layout"
  data-divider={undefined}
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
  <div
    className="zen_ui__horizontal_layout_column"
    style={Object {}}
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Horizontal Layout should render correctly 3`] = `
<div
  className="zen_ui__horizontal_layout"
  data-divider={undefined}
  style={
    Object {
      "height": "100%",
    }
  }
>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;

exports[`Horizontal Layout should render correctly 4`] = `
<div
  className="zen_ui__horizontal_layout"
  data-divider={undefined}
  style={Object {}}
>
  <div
    className="zen_ui__horizontal_layout_column"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      style={
        Object {
          "padding": null,
          "width": null,
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;
